ganttproject/src/main/java/net/sourceforge/ganttproject/GanttProject.java

```diff
@@ -956,33 +951,62 @@ public void refresh() {
     super.repaint();
   }
 
-  // TODO: Accumulate changes instead of sending it every time.
+  private interface TxnSendListener {
+    void onSendCompleted();
+  }
+
+  private final AtomicReference<TxnSendListener> txnSendingListener = new AtomicReference<>();
+
   private Unit sendProjectStateLogs() {
     gpLogger.debug("Sending project state logs");
+    if (txnSendingListener.get() != null) return Unit.INSTANCE;
     try {
       var baseTxnCommitInfo = myBaseTxnCommitInfo.get();
       var txns = myProjectDatabase.fetchTransactions(baseTxnCommitInfo.right + 1, 1);
       if (!txns.isEmpty()) {
-        getWebSocket().sendLogs(new InputXlog(
-          baseTxnCommitInfo.left,
-          "userId",
-          "refid",
-          txns
-        ));
+        var listener = new TxnSendListener() {
+          @Override
+          public void onSendCompleted() {
+            txnSendingListener.compareAndSet(this, null);
+          }
+        };
+        if (txnSendingListener.compareAndSet(null, listener)) {
+          getWebSocket().sendLogs(new InputXlog(
+            baseTxnCommitInfo.left,
+            "userId",
+            "refid",
+            txns
+          ));
+        }
       }
     } catch (ProjectDatabaseException e) {
       gpLogger.error("Failed to send logs", new Object[]{}, ImmutableMap.of(), e);
```

Comment:
If we are sure that `ProjectDatabaseException` will not be thrown when listener is set then maybe move the body of `if` out of try?

Reply:
Sounds good, done